### PR TITLE
Added Makefile

### PR DESCRIPTION
**Pull request by jpkrohling**
Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

**Comment by LiliC** [comment]
Yes please! :) This was very helpful, when I looked how to contribute 👍 

cc @metalmatze @kakkoyun

**Comment by kakkoyun** [comment]
lgtm, @metalmatze should take a look at it, since he has already started a discussion around it at #64.

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,42 @@
+JSONNET_FILES=$(shell find . -type f -not -path './vendor/*' \( -name '*.libsonnet' -o -name '*.jsonnet' \))
+
+.DEFAULT_GOAL := all
+
+.PHONY: check
+check:
+	@echo Checking...
+	@echo $(JSONNET_FILES) | xargs -L1 jsonnetfmt -i
+	@git diff -s --exit-code . || (echo "Build failed: one of more files aren't staged to be committed." && git diff && exit 1)
```
Comment (metalmatze) [inline Makefile]: As commented in an earlier PR, can we not suppress these diffs, I'd like to see what's going on.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	@rm -rf environments/kubernetes/manifests
+	@rm -rf environments/openshift/manifests
+	@rm -rf environments/sre/servicemonitors
+	@rm -rf environments/sre/prometheusrules
+	@rm -rf environments/sre/grafana
+
+.PHONY: build
+build:
+	@echo Building...
+	@./build.sh
```
Comment (metalmatze) [inline Makefile]: If we want this Makefile we should replace the `build.sh` with actual make targets like:

```Makefile
environments/kubernetes/manifests:
jsonnet -J vendor -m environments/kubernetes/manifests environments/kubernetes/main.jsonnet | xargs -I{} sh -c 'cat {} | gojsontoyaml > {}.yaml; rm -f {}' -- {}

# for the rest of the build.sh the same
...
```

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+.PHONY: vendor
+vendor:
+	@echo Bringing dependencies...
+	@jb install
```
Comment (metalmatze) [inline Makefile]: We currently don't suppress the output of jb anywhere.

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	@go get \
+		github.com/google/go-jsonnet/cmd/jsonnet \
+		github.com/jsonnet-bundler/jsonnet-bundler/cmd/jb \
+		github.com/brancz/gojsontoyaml
+
+.PHONY: all
+all: check vendor build
+
+.PHONY: ci
+ci: all check # we run the check twice, as the manifests might have changed as part of the build
```
Comment (metalmatze) [inline Makefile]: We usually check at the end of running everything. I don't really get why we would check at the beginning? :thinking:

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+.PHONY: vendor
+vendor:
+	@echo Bringing dependencies...
+	@jb install
+
+.PHONY: install-tools
+install-tools:
+	@## this target is missing the installation of jsonnetfmt, which doesn't seem to have a binary available (!!)
+	@go get \
```
Comment (metalmatze) [inline Makefile]: We could also check of the binaries being available and always run this target when simply running `make`.

File: Makefile (hunk 1 above)
Comment (jpkrohling) [reply Makefile]: It is doing that, but only if it fails. Look at the `&& git diff && exit 1` at the end of the command.

File: Makefile (hunk 4 above)
Comment (jpkrohling) [reply Makefile]: Fail fast :)

File: Makefile (hunk 2 above)
Comment (jpkrohling) [reply Makefile]: I'd rather keep the Makefile as a wrapper for more complex scripts. When we have simple commands, like `jb install`, it's OK to have it here, but once it gets more complex, it's easier to maintain shell scripts than make targets.

**Comment by kakkoyun** [comment]
@metalmatze @jpkrohling Shall we push this through? I could help with the fixes if you don't have time.

**Comment by jpkrohling** [comment]
From my side: yes, please!

**Comment by jpkrohling** [comment]
Closing, as I don't think this will be merged anytime soon. Not in this form, at least :-)

**Pull request closed without merging**